src/include/server/mir/input/input_event_transformer.h

```diff
@@ -0,0 +1,113 @@
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_INPUT_INPUT_EVENT_TRANSFORMER_H_
+#define MIR_INPUT_INPUT_EVENT_TRANSFORMER_H_
+
+#include "mir/input/event_filter.h"
+#include "mir/input/event_builder.h"
+
+#include <chrono>
+#include <mutex>
+#include <vector>
+#include <memory>
+
+namespace mir
+{
+class MainLoop;
+namespace input
+{
+class VirtualInputDevice;
+class InputDeviceRegistry;
+class InputSink;
+class EventBuilder;
+class InputEventTransformer : public EventFilter
+{
+public:
+    class EventDispatcher
+    {
+    public:
+        EventDispatcher(
+            std::shared_ptr<MainLoop> const main_loop,
+            input::InputSink* const sink,
+            input::EventBuilder* const builder);
+
+        void dispatch_key_event(
+            std::optional<std::chrono::nanoseconds> timestamp,
+            MirKeyboardAction action,
+            xkb_keysym_t keysym,
+            int scan_code);
+
+        void dispatch_pointer_event(
+            std::optional<std::chrono::nanoseconds> timestamp,
+            MirPointerAction action,
+            MirPointerButtons buttons,
+            std::optional<mir::geometry::PointF> position,
+            mir::geometry::DisplacementF motion,
+            MirPointerAxisSource axis_source,
+            events::ScrollAxisH h_scroll,
+            events::ScrollAxisV v_scroll);
+
+        void maybe_update(InputSink* maybe_new_sink, EventBuilder* maybe_new_builder);
+
+    private:
+        void dispatch(mir::EventUPtr e);
+
+        std::shared_ptr<MainLoop> const main_loop;
+
+        // Since dispatchers are handed down to transformers which might keep a
+        // reference to them, then we should be careful to not change the
+        // builder as they dispatch an event.
+        std::mutex mutex;
+
+        // Sinks and/or builders are not constant for one virtual device.
+        input::InputSink* sink;
+        input::EventBuilder* builder;
+    };
+
+    class Transformer
+    {
+    public:
+        virtual ~Transformer() = default;
+
+        // Returning true means that the event has been successfully processed and
+        // shouldn't be handled by later transformers, whether the transformer is
+        // accumulating events for later dispatching or has immediately dispatched
+        // an event is an implementation detail of the transformer
+        virtual bool transform_input_event(std::shared_ptr<EventDispatcher> const& dispatcher,  MirEvent const&) = 0;
```

Comment:
I am of the opinion that this would be easier without the need for the `EventDispatcher` class at all, especially since the event builder class will have to keep in sync with anything that could be built (e.g. touch events). 

```suggestion
        virtual bool transform_input_event(MirEvent const&, input::InputSink* sink, input::EventBuilder* builder, std::function<void(mir::EventUPtr e) const& dispatch) = 0;
```

I think this gets rid of some unnecessary complexity. What do you think?

Reply:
I originally added `EventDispatcher` to hoist the issuing of the even onto another thread (the whole purpose of `main_loop` in the struct). You can see for yourself, just apply these changes and the first movement should occur, but then miral should deadlock.

```diff
diff --git a/src/include/server/mir/input/input_event_transformer.h b/src/include/server/mir/input/input_event_transformer.h
index 2153dc90d1..801821a425 100644
--- a/src/include/server/mir/input/input_event_transformer.h
+++ b/src/include/server/mir/input/input_event_transformer.h
@@ -63,7 +63,6 @@ public:
 
         void maybe_update(InputSink* maybe_new_sink, EventBuilder* maybe_new_builder);
 
-    private:
         void dispatch(mir::EventUPtr e);
 
         std::shared_ptr<MainLoop> const main_loop;
diff --git a/src/server/shell/accessibility_manager.cpp b/src/server/shell/accessibility_manager.cpp
index f116f21866..c15482ed0e 100644
--- a/src/server/shell/accessibility_manager.cpp
+++ b/src/server/shell/accessibility_manager.cpp
@@ -100,7 +100,7 @@ struct MouseKeysTransformer: public mir::input::InputEventTransformer::Transform
                 return false;
             }
 
-            dispatcher->dispatch_pointer_event(
+            dispatcher->sink->handle_input(dispatcher->builder->pointer_event(
                 std::nullopt,
                 mir_pointer_action_motion,
                 0,
@@ -108,7 +108,7 @@ struct MouseKeysTransformer: public mir::input::InputEventTransformer::Transform
                 offset,
                 mir_pointer_axis_source_none,
                 mir::events::ScrollAxisH{},
-                mir::events::ScrollAxisV{});
+                mir::events::ScrollAxisV{}));
 
             return true;
         }
```

Reply:
I think I misspoke when I said you should pass the `input_sink` :)

Still, you could just pass the `builder` and a `dispatch` function as parameters to this callback and avoid the `EventDispatcher` altogether. 

e.g.

```c++
// input_event_transformer.cpp
bool mi::InputEventTransformer::handle(MirEvent const& event)
{
    ...
    auto dispatch = [sink=sink, main_loop=main_loop](mir::EventUPtr e) {
        main_loop->spawn(
            [e = std::shared_ptr(std::move(e)), this]
            {
                sink->handle_input(e);
            });
    };

    if (t->transform_input_event(event, builder, dispatch)) { ... }
}
```

The builder part of the event does not have to be thread safe. Only the dispatch part does. Hence, we can remove a lot of complexity by capturing it in a lambda. Then, we can do:

```c++
// accessibility_manager.cpp

dispatch(builder->pointer_event(...))
```

Reply:
In the context of #3757, this might break since we'll capture the sink and builder and keep them as long as the user holds one of the keys used to move the mouse

Reply:
I don't understand how it is any different from what you have however, since it just removes the class but keeps all of the same functionality. Perhaps I'm missing something?

Reply:
`std::shared_ptr<EventDispatcher>` adds an extra level of indirection that should allow the pointers to be updated when they change. Now, I investigate a little more and the pointers should not change, so I'm very confused at the moment